# 프로젝트 1 손글씨를 분류해보자.

라이브러리 버전을 확인하자


In [130]:
import sklearn

print(sklearn.__version__)

1.1.2


사이킷런에서 필요한 모듈들을 임포트한다.

In [131]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

## load_digits 메서드로 데이터를 준비하자

In [138]:
handscript = load_digits()
print(handscript.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

## 데이터 이해하기
1. Feature Data 지정하기
2. Label Data 지정하기
3. Target Names 출력해보기
4. 데이터 Describe 해보기


In [152]:
print(handscript.keys())

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])


In [156]:
print(handscript.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [146]:
print(handscript.data)

[[ 0.  0.  5. ...  0.  0.  0.]
 [ 0.  0.  0. ... 10.  0.  0.]
 [ 0.  0.  0. ... 16.  9.  0.]
 ...
 [ 0.  0.  1. ...  6.  0.  0.]
 [ 0.  0.  2. ... 12.  0.  0.]
 [ 0.  0. 10. ... 12.  1.  0.]]


손글씨 데이터의 값은 8*8의 16레벨의 값이다.

In [159]:
print(handscript.target)
print(len(handscript.target))
print(handscript.target[:20])

[0 1 2 ... 8 9 8]
1797
[0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9]


타겟은 1797개의 샘플이 0~9까지의 값을 갖고 있다.

In [151]:
handscript.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

타겟의 이름은 0~9의 숫자이다.

그럼 이제 trainset과 testset을 분리해보자.

In [161]:
X_train, X_test, y_train, y_test = train_test_split(handscript.data, handscript.target, test_size = 0.2, random_state = 32)

In [162]:
print(len(X_train), len(X_test))

1437 360


# 다양한 모델로 학습시켜보자.
1. Decision Tree
2. Random Forest
3. SVM
4. SGD Classifier
5. Logistic regression


### Decision tree for handscipt

In [183]:
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier(random_state=32)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

print(classification_report(y_test, y_pred))
confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        38
           1       0.75      0.83      0.79        36
           2       0.75      0.84      0.79        32
           3       0.92      0.86      0.89        56
           4       0.85      0.90      0.88        31
           5       0.92      0.97      0.95        36
           6       0.97      0.94      0.96        34
           7       0.94      0.88      0.91        34
           8       0.88      0.78      0.82        27
           9       0.79      0.83      0.81        36

    accuracy                           0.88       360
   macro avg       0.88      0.87      0.87       360
weighted avg       0.88      0.88      0.88       360



0.875

### Random Forest


In [184]:
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(random_state=32)
rf_clf.fit(X_train, y_train)
y_pred = rf_clf.predict(X_test)

print(classification_report(y_test, y_pred))
confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        38
           1       0.97      1.00      0.99        36
           2       0.97      1.00      0.98        32
           3       1.00      1.00      1.00        56
           4       1.00      0.97      0.98        31
           5       1.00      0.97      0.99        36
           6       1.00      1.00      1.00        34
           7       0.97      1.00      0.99        34
           8       0.96      0.89      0.92        27
           9       0.97      1.00      0.99        36

    accuracy                           0.99       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.99      0.99      0.99       360



0.9861111111111112

### SVM


In [187]:
from sklearn.svm import SVC

svm_clf = SVC(random_state=32)
svm_clf.fit(X_train,y_train)
y_pred = svm_clf.predict(X_test)

print(classification_report(y_test,y_pred))
confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        38
           1       1.00      1.00      1.00        36
           2       1.00      1.00      1.00        32
           3       1.00      1.00      1.00        56
           4       1.00      0.97      0.98        31
           5       1.00      0.97      0.99        36
           6       1.00      1.00      1.00        34
           7       1.00      1.00      1.00        34
           8       1.00      1.00      1.00        27
           9       0.95      1.00      0.97        36

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



0.9944444444444445

### SGD Classifier


In [188]:
from sklearn.linear_model import SGDClassifier

sgd_clf = SGDClassifier(random_state=32)
sgd_clf.fit(X_train,y_train)
y_pred = sgd_clf.predict(X_test)

print(classification_report(y_test, y_pred))
confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        38
           1       1.00      0.78      0.88        36
           2       1.00      0.97      0.98        32
           3       0.98      0.95      0.96        56
           4       1.00      0.97      0.98        31
           5       0.97      0.94      0.96        36
           6       1.00      1.00      1.00        34
           7       1.00      0.97      0.99        34
           8       0.75      1.00      0.86        27
           9       0.86      1.00      0.92        36

    accuracy                           0.95       360
   macro avg       0.96      0.95      0.95       360
weighted avg       0.96      0.95      0.95       360



0.9527777777777777

### Logistic Regression


In [189]:
from sklearn.linear_model import LogisticRegression

lr_clf = LogisticRegression(random_state=32,max_iter=50)
lr_clf.fit(X_train, y_train)
y_pred = lr_clf.predict(X_test)

print(classification_report(y_test, y_pred))
confusion_matrix(y_test, y_pred)
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       1.00      0.89      0.94        38
           1       0.95      0.97      0.96        36
           2       1.00      1.00      1.00        32
           3       1.00      0.98      0.99        56
           4       1.00      0.97      0.98        31
           5       0.89      0.94      0.92        36
           6       1.00      1.00      1.00        34
           7       1.00      1.00      1.00        34
           8       0.96      0.96      0.96        27
           9       0.92      1.00      0.96        36

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



/Users/taehwan/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.9722222222222222

# 모델을 평가해보기

sklearn.metrics에서 제공하는
classfication_report와 confusion_matrix, accuracy_score 세 가지 평가지표를 확인해봤는데
분류 숫자가 많아서 matrix는 보기 어렵다.
숫자를 인식해서 출력해야하는데, 최대한 많은 수를 정확하게 인식해서 출력해주는 것이 중요하기에 accuracy_score를 보는 것이 좋겠다.
가장 높은 svm의 방식이 가장 우수해보인다.

또한 개별적인 숫자의 정확도를 확인하기 위해서는 Classification_report를 확인하는 것이 좋겠다.
svm모델의 경우, report로 자세히 보면 4,5에서 f1_score가 적게 나오고 있다. 그러면 4,5에 인풋데이터를 늘리거나 지도학습을 강화하거나 할 수 있겠다.

그리고 오류가 어느 숫자로 판별해 FN이 되었는지 확인하기 위해서는 matrix를 보면 좋겠다.


# 프로젝트2 와인을 분류해봅시다.


In [195]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

## 데이터 준비하기
load_wine

In [196]:
wineset = load_wine()

## 데이터 이해하기

In [197]:
X = wineset.data
y = wineset.target
print(wineset.target_names)
wineset.DESCR

['class_0' 'class_1' 'class_2']


'.. _wine_dataset:\n\nWine recognition dataset\n------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 178\n    :Number of Attributes: 13 numeric, predictive attributes and the class\n    :Attribute Information:\n \t\t- Alcohol\n \t\t- Malic acid\n \t\t- Ash\n\t\t- Alcalinity of ash  \n \t\t- Magnesium\n\t\t- Total phenols\n \t\t- Flavanoids\n \t\t- Nonflavanoid phenols\n \t\t- Proanthocyanins\n\t\t- Color intensity\n \t\t- Hue\n \t\t- OD280/OD315 of diluted wines\n \t\t- Proline\n\n    - class:\n            - class_0\n            - class_1\n            - class_2\n\t\t\n    :Summary Statistics:\n    \n    ============================= ==== ===== ======= =====\n                                   Min   Max   Mean     SD\n    ============================= ==== ===== ======= =====\n    Alcohol:                      11.0  14.8    13.0   0.8\n    Malic Acid:                   0.74  5.80    2.34  1.12\n    Ash:                          1.36  3.23    2.36  0.27\

train, test 데이터 분리

In [199]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=32)

### Decision tree


In [206]:
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier(random_state=32)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       0.94      1.00      0.97        16
           1       0.89      0.80      0.84        10
           2       0.90      0.90      0.90        10

    accuracy                           0.92        36
   macro avg       0.91      0.90      0.90        36
weighted avg       0.92      0.92      0.91        36

[[16  0  0]
 [ 1  8  1]
 [ 0  1  9]]


0.9166666666666666

### Random forest


In [217]:
from sklearn.ensemble import RandomForestClassifier


rf_clf = RandomForestClassifier(random_state=32)
rf_clf.fit(X_train, y_train)
y_pred = rf_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       1.00      0.90      0.95        10
           2       0.91      1.00      0.95        10

    accuracy                           0.97        36
   macro avg       0.97      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36

[[16  0  0]
 [ 0  9  1]
 [ 0  0 10]]


0.9722222222222222

### SVM


In [235]:
from sklearn.svm import SVC


X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=5)

svm_clf = SVC(random_state=5)
svm_clf.fit(X_train, y_train)
y_pred = svm_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       0.86      0.71      0.77        17
           1       0.69      0.82      0.75        11
           2       0.44      0.50      0.47         8

    accuracy                           0.69        36
   macro avg       0.66      0.67      0.66        36
weighted avg       0.72      0.69      0.70        36

[[12  0  5]
 [ 2  9  0]
 [ 0  4  4]]


0.6944444444444444

### SGD Classifier


In [242]:
from sklearn.linear_model import SGDClassifier

sgd_clf = SGDClassifier(random_state=3)
sgd_clf.fit(X_train,y_train)
y_pred = sgd_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       0.76      0.94      0.84        17
           1       0.80      0.36      0.50        11
           2       0.40      0.50      0.44         8

    accuracy                           0.67        36
   macro avg       0.65      0.60      0.60        36
weighted avg       0.69      0.67      0.65        36

[[16  0  1]
 [ 2  4  5]
 [ 3  1  4]]


0.6666666666666666

### Logistic Regression


In [212]:
from sklearn.linear_model import LogisticRegression

lr_clf = LogisticRegression(random_state =32, max_iter=300)
lr_clf.fit(X_train, y_train)
y_pred = lr_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       0.90      0.90      0.90        10
           2       0.90      0.90      0.90        10

    accuracy                           0.94        36
   macro avg       0.93      0.93      0.93        36
weighted avg       0.94      0.94      0.94        36

[[16  0  0]
 [ 0  9  1]
 [ 0  1  9]]


/Users/taehwan/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.9444444444444444

## 모델 평가하기

SVM 모델과 SGD 모델이 3가지로 분류해내지 못했다.

random_state의 값을 바꾸면 다시 분류가 되기도한다. 완전하게 이해되진 않지만
적은 샘플 수때문에 이런 일이 발생하지 않나 싶다.

우선 SVM과 SGD에서는 적은 샘플에 적합하지 않아 보인다.

# 프로젝트3, 유방암 진단

In [245]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

### 데이터 준비


In [246]:
breast_cancer = load_breast_cancer()

### 데이터 이해하기


In [247]:
X = breast_cancer.data
y = breast_cancer.target
print(breast_cancer.target_names)
print(breast_cancer.DESCR)

['malignant' 'benign']
.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instanc

### train, test 데이터 분리


In [249]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=32)

### 다양한 모델로 학습시켜보기



### Decision Tree

In [251]:
from sklearn.tree import DecisionTreeClassifier

dt_clf = DecisionTreeClassifier(random_state=32)
dt_clf.fit(X_train,y_train)
y_pred = dt_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       0.84      0.84      0.84        44
           1       0.90      0.90      0.90        70

    accuracy                           0.88       114
   macro avg       0.87      0.87      0.87       114
weighted avg       0.88      0.88      0.88       114

[[37  7]
 [ 7 63]]


0.8771929824561403

### Random Forest

In [253]:
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(random_state=32)
rf_clf.fit(X_train, y_train)
y_pred = rf_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       0.91      0.93      0.92        44
           1       0.96      0.94      0.95        70

    accuracy                           0.94       114
   macro avg       0.93      0.94      0.94       114
weighted avg       0.94      0.94      0.94       114

[[41  3]
 [ 4 66]]


0.9385964912280702

### SVM

In [255]:
from sklearn.svm import SVC

svm_clf = SVC(random_state=32)
svm_clf.fit(X_train,y_train)
y_pred = svm_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       0.91      0.73      0.81        44
           1       0.85      0.96      0.90        70

    accuracy                           0.87       114
   macro avg       0.88      0.84      0.85       114
weighted avg       0.87      0.87      0.86       114

[[32 12]
 [ 3 67]]


0.868421052631579

### SGD Classifier

In [257]:
from sklearn.linear_model import SGDClassifier

sgd_clf = SGDClassifier(random_state=32)
sgd_clf.fit(X_train, y_train)
y_pred = sgd_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)

              precision    recall  f1-score   support

           0       0.87      0.89      0.88        44
           1       0.93      0.91      0.92        70

    accuracy                           0.90       114
   macro avg       0.90      0.90      0.90       114
weighted avg       0.90      0.90      0.90       114

[[39  5]
 [ 6 64]]


0.9035087719298246

### Logistic Regression

In [265]:
from sklearn.linear_model import LogisticRegression

lg_clf = LogisticRegression(random_state = 32, max_iter=300)
lg_clf.fit(X_train, y_train)
y_pred = lg_clf.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
accuracy_score(y_test, y_pred)


              precision    recall  f1-score   support

           0       0.89      0.89      0.89        44
           1       0.93      0.93      0.93        70

    accuracy                           0.91       114
   macro avg       0.91      0.91      0.91       114
weighted avg       0.91      0.91      0.91       114

[[39  5]
 [ 5 65]]


/Users/taehwan/miniconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.9122807017543859

## 모델평가



Acurracy_score는 랜덤포레스트 모델이 93점 대로 가장 높았다.

유방암은 악성(class_0)을 양성(class_1)으로 판단하지 않도록 최적화 하는 것이 중요하다.
따라서 precision보다 class_0의 recall이 높은 값이여야한다.

class_0의 recall 값
decision_tree 0.84
random_forest 0.93
svm 0.73
sgd 0.89
logistic_regression 0.89

랜덤포레스트가 정확도 및 class_0의 recall 모두 최우수였기에 이번 데이터 분석에 효과적인 모델이라고 볼 수 있다.

